core/shared/src/main/scala/fs2/internal/Resource.scala

```diff
@@ -0,0 +1,165 @@
+  *
+  * (1) The Algebra.Resource is interpreted. In this case resource finalization will be invoked if
+  *     the resource was successfully acquired (the `acquired` was evaluated and resource was not leased other scope)
+  * (2) The owning Scope was closed by `Algebra.CloseScope`. This essentially evaluates `release` of
+  *     the Resource and acts like (1).
+  * (3) The `acquired` was evaluated after scope was `released` by either (1) or (2). In this case
+  *     resource will invoke the finalization immediately if the resource is not leased.
+  * (4) The `cancelLEase` is invoked for previously leased resource. This will invoke resource
+  *     finalization if resource was already acquired and released and there are no more outstanding leases.
+  *
+  * Resources may be leased to other scopes. in that case, each scope must lease resource with `lease` and
+  * when scope is closed (or when the resource lease is no longer required) release the lease with `cancelLease`.
+  *
+  * Note thate every method which as part of its invocation may potentially call resource finalizer
+  * is returning `F[Either[Throwable, Unit]]` instead just simple F[Unit] to make sure the errors when releasing resource
+  * are properly handled.
+  *
+  *
+  */
+private[internal] sealed trait Resource[F[_]] {
+
+  /** if of the resource **/
+  def id: Token
+
+  /**
+    * Depending on resource state this will either release resource, or when resource was not yet fully
+    * acquired, this will schedule releasing of the resource at earliest opportunity, that is when:
+    *
+    * (a) The resource finished its acquisition
+    *  - and -
+    * (b) All scopes borrowing this resource released resource.
+    *
+    * As a part of release process the resoure's finalizer may be run and in that case it may fail eventually.
+    *
+    * @return
+    */
+  def release: F[Either[Throwable, Unit]]
+
+  /**
+    * Signals that resource was succesfully acquired. Finalizer to be run on release is provided.
+    *
+    * If the resource was closed before beqing acquired, then supplied finalizer is run.
+    * That may result in finalizer eventually failing.
+    *
+    * @param finalizer
+    * @return
+    */
+  def acquired(finalizer: F[Unit]): F[Either[Throwable, Unit]]
+
+  /**
+    * Signals that this resource was leased by another scope than one allocating this resource.
+    *
+    * Yields to true, if this resource was successfully leased, and scope must `cancelLease` it
+    * or to `false` when this resource cannot be leased because resource is already released.
+    */
+  def lease: F[Boolean]
+
+  /**
+    * Each leased resource must be canceled when no longer needed by scope that leased it.
+    *
+    * When the lease is returned, this may result in resource's finalizer to be invoked and
+    * thus this may result in finalizer to fail.
+    *
+    */
+  def cancelLease: F[Either[Throwable, Unit]]
```

Comment:
This might not be a good idea but we could encode the lease state machine directly in the types by changing `lease` to be `def lease: F[Option[Lease[F]]]` where:
```scala
trait Lease[F[_]] {
  def cancel: F[Either[Throwable,Unit]]
}
```

Reply:
perhaps you right just lease: F[Lease[F]]. Not sure if we really need that option here.